README.md
My Bookmarklets v1.3.1
======================

Repository of the bookmarklets I've used over the years


Markdown Links
--------------

* Now YouTube aware capturing channel name and video duration!
* Now Rust community (crates.io, docs.rs, and libs.rs) aware!
	* Awareness in this case indicates normalization of titles to make my OCD happy



### Markdown Page Link

`markdown-page-link`

Opens an alert with the `[PAGE_TITLE](URL)` syntax for the active page


### Markdown Page Ref

`markdown-page-ref`

Opens an alert with the `[PAGE_TITLE]: URL` syntax for the active page


### Markdown Selection Link

`markdown-selection-link`

Opens an alert with the `[SELECTION - PAGE_TITLE](URL)` syntax for the active page


### Markdown Selection Ref

`markdown-selection-link`

Opens an alert with the `[SELECTION - PAGE_TITLE]: URL` syntax for the active page


Resources
---------

* [Bookmarklet Maker][]: great for testing and minification for deployment



[Bookmarklet Maker]: https://caiorss.github.io/bookmarklet-maker/



CHANGELOG.md
ChangeLog
=========


Version 1.3.1
-------------

2023-10-22 - 2023-10-24

* Updated RegEx for recent YouTube changes
* Added yesterdays changes to the selection versions as well (DOH!)
* Added code to handle when the tooltip date is not available


Version 1.3.0
-------------

2023-10-21

* Converted to custom parseDate function to hand relative dates such as "Premiere 3 days ago"


Version 1.2.1
-------------

2023-04-09

* Updated YouTube awareness to not fail on non-watch pages


Version 1.2.0
-------------

2023-04-05

* Added CHANGELOG.md
* Added markdown-selection-ref.js
* Added markdown-selection-ref-min.js
* Updated all scripts to be Rust community (crates.io, docs.rs, and libs.rs) aware




markdown-page-ref.js
/*
M↓ Page Ref

Version: 1.3.1
Description: YouTube and Rust community aware
Updated: 2023-10-24
*/
javascript:(
	() => {
		let D=document, title=D.title;

		function parseDate(date) {
			// console.log(`parseDate() | > input: "${date}" | test: ${checkDateRE.test(date)}`);
			const checkDateRE = /^(?<prefix>[Pp]remiered?)? ?((?<count>\d+) (?<unit>\w+) (ago)|(?<month>\w+) (?<day>\d+), (?<year>\d+))$/;
			let result = '';

			if (checkDateRE.test(date)) {
				let {prefix, count, unit, month, day, year} = checkDateRE.exec(date).groups;
				let newDate = new Date(Date.now());
				let transDay = new Date(newDate.getFullYear(), newDate.getMonth(), newDate.getDate());
				let transMonth = new Date(newDate.getFullYear(), newDate.getMonth());
				let transYear = new Date(newDate.getFullYear(), 0, 1);

				// console.log(`parseDate() |  prefix: ${prefix} | count: ${count} | unit: ${unit} | month: ${month} | day: ${day} | year: ${year}`);

				if (year != null && month != null && day != null) {
					console.log(`parseDate() | Input Date Normal | ${date}`);
					newDate = new Date(date);
					result = newDate.toISOString().substr(0, 10);
				} else if (count != null && unit != null) {
					let days = 0;

					console.log(`parseDate() | Input Type Relative | ${date}`);
					unit = unit.toLowerCase();
					console.log(`parseDate() |    unit: ${unit}`);
					switch (unit) {
					case 'second':
					case 'seconds':
						if (newDate.getTime() < transYear.getTime()) {
							newDate.setFullYear(newDate.getFullYear() - 1);
						}
						// break;
					case 'minute':
					case 'minutes':
					case 'hour':
					case 'hours':
						// Our job is done
						result = newDate.toISOString().substr(0, 10)
						break;
					case 'day':
					case 'days':
					case 'week':
					case 'weeks':
						newDate.setHours(0);
						newDate.setMinutes(0);
						newDate.setSeconds(0);
						if (unit === 'week' || unit === 'weeks') {
							count = count * 7;
						}
						// console.log(`parseDate() | 0 | newDate.toISOString().substr(0, 10): ${newDate.toISOString().substr(0, 10)}`);
						days = newDate.getDate();
						// console.log(`parseDate() | 0 | days: ${days} | count: ${count}`);
						if (count < days) {
							days -= count;
							// console.log(`parseDate() | 1 | days: ${days} | count: ${count}`);
							newDate.setDate(days);
							// console.log(`parseDate() | 2 | days: ${newDate.getDate()} | count: ${count}`);
						} else {
							let mon = newDate.getMonth();
							if (mon === 0) {
								newDate.setYear(newDate.getFullYear() - 1);
								newDate.setMonth(11);
							} else {
								newDate.setMonth(mon - 1);
							}
							// console.log(`parseDate() |     mon: ${mon}`);
						}
						if (unit === 'week' || unit === 'weeks') {
							result = newDate.toISOString().substr(0, 7)
						} else {
							result = newDate.toISOString().substr(0, 10)
						}
						// console.log(`parseDate() | 3 | days: ${newDate.getDate()} | result: ${result}`);
						break;
					case 'month':
					case 'months':
						newDate.setDate(1)
						newDate.setHours(0);
						newDate.setMinutes(0);
						newDate.setSeconds(0);

						let mon = newDate.getMonth();
						console.log(`parseDate() |     mon: ${mon} | count: ${count}`);
						if (count < mon) {
							newDate.setMonth(mon - count);
						} else {
							newDate.setFullYear(newDate.getFullYear() - 1);
							mon = 12 + (mon - count);
							// console.log(`parseDate() |     mon: ${mon}`);
							newDate.setMonth(mon);
						}
						result = newDate.toISOString().substr(0, 7)
						console.log(`parseDate() | month | result: ${result}`);
						break;
					case 'year':
					case 'years':
						result = newDate.getFullYear() - count;
						break;
					default:
						console.error(`parseDate() | unknown unit: '${unit}'`)
					}
				}
				// console.log(`parseDate() |   count: ${count} | unit: ${unit} | year: ${year} | month: ${month} | day: ${day}`);
				// console.log(`parseDate() | newDate: ${newDate.toISOString().substr(0, 10)}`);
				// console.log(`parseDate() |  result: ${result}`);
			}
			// console.log(`parseDate() | result: ${result} | input: "${date}" | test: ${checkDateRE.test(date)}`);
			return result
		}

		switch(true) {
		case D.URL.indexOf('crates.io') !== -1:
			title = D.title.replace('crates.io: Rust Package Registry', 'Crates.io');
			break;
		case D.URL.indexOf('docs.rs') !== -1:
			title = D.title.replace('Rust', 'Docs.rs');
			break;
		case D.URL.indexOf('lib.rs') !== -1:
			title = D.title.replace(' // ', ' - ').replace('—', '-');
			break;
		case D.URL.indexOf('youtube.com/watch') !== -1:
			let channel = D.querySelector('div#owner #text > a').innerText;
			let duration = D.querySelector('#movie_player .ytp-time-duration').innerText;
			let posted = D.querySelector('#description #tooltip').innerText.trim();
			if (posted !== '') {
				posted = posted.split('•')[1].trim();
			} else {
				posted = D.querySelector('#description #info span:nth-child(3)').innerText.trim();
			}
			posted = parseDate(posted);
			title = D.title.replace('YouTube',`${channel} - ${posted} - ${duration} - YouTube`);
			break;
		};

		alert(`[${title.replaceAll('[', "\\[").replaceAll(']', "\\]").replaceAll('|', '&vert;')}]: ${D.URL}`);
	}
)();

markdown-page-link.js
/*
M↓ Page Link

Version: 1.3.1
Description: YouTube and Rust community aware
Updated: 2023-10-24
*/
javascript:(
	() => {
		let D=document, title=D.title;

		function parseDate(date) {
			// console.log(`parseDate() | > input: "${date}" | test: ${checkDateRE.test(date)}`);
			const checkDateRE = /^(?<prefix>[Pp]remiered?)? ?((?<count>\d+) (?<unit>\w+) (ago)|(?<month>\w+) (?<day>\d+), (?<year>\d+))$/;
			let result = '';

			if (checkDateRE.test(date)) {
				let {prefix, count, unit, month, day, year} = checkDateRE.exec(date).groups;
				let newDate = new Date(Date.now());
				let transDay = new Date(newDate.getFullYear(), newDate.getMonth(), newDate.getDate());
				let transMonth = new Date(newDate.getFullYear(), newDate.getMonth());
				let transYear = new Date(newDate.getFullYear(), 0, 1);

				// console.log(`parseDate() |  prefix: ${prefix} | count: ${count} | unit: ${unit} | month: ${month} | day: ${day} | year: ${year}`);

				if (year != null && month != null && day != null) {
					console.log(`parseDate() | Input Date Normal | ${date}`);
					newDate = new Date(date);
					result = newDate.toISOString().substr(0, 10);
				} else if (count != null && unit != null) {
					let days = 0;

					console.log(`parseDate() | Input Type Relative | ${date}`);
					unit = unit.toLowerCase();
					console.log(`parseDate() |    unit: ${unit}`);
					switch (unit) {
					case 'second':
					case 'seconds':
						if (newDate.getTime() < transYear.getTime()) {
							newDate.setFullYear(newDate.getFullYear() - 1);
						}
						// break;
					case 'minute':
					case 'minutes':
					case 'hour':
					case 'hours':
						// Our job is done
						result = newDate.toISOString().substr(0, 10)
						break;
					case 'day':
					case 'days':
					case 'week':
					case 'weeks':
						newDate.setHours(0);
						newDate.setMinutes(0);
						newDate.setSeconds(0);
						if (unit === 'week' || unit === 'weeks') {
							count = count * 7;
						}
						// console.log(`parseDate() | 0 | newDate.toISOString().substr(0, 10): ${newDate.toISOString().substr(0, 10)}`);
						days = newDate.getDate();
						// console.log(`parseDate() | 0 | days: ${days} | count: ${count}`);
						if (count < days) {
							days -= count;
							// console.log(`parseDate() | 1 | days: ${days} | count: ${count}`);
							newDate.setDate(days);
							// console.log(`parseDate() | 2 | days: ${newDate.getDate()} | count: ${count}`);
						} else {
							let mon = newDate.getMonth();
							if (mon === 0) {
								newDate.setYear(newDate.getFullYear() - 1);
								newDate.setMonth(11);
							} else {
								newDate.setMonth(mon - 1);
							}
							// console.log(`parseDate() |     mon: ${mon}`);
						}
						if (unit === 'week' || unit === 'weeks') {
							result = newDate.toISOString().substr(0, 7)
						} else {
							result = newDate.toISOString().substr(0, 10)
						}
						// console.log(`parseDate() | 3 | days: ${newDate.getDate()} | result: ${result}`);
						break;
					case 'month':
					case 'months':
						newDate.setDate(1)
						newDate.setHours(0);
						newDate.setMinutes(0);
						newDate.setSeconds(0);

						let mon = newDate.getMonth();
						console.log(`parseDate() |     mon: ${mon} | count: ${count}`);
						if (count < mon) {
							newDate.setMonth(mon - count);
						} else {
							newDate.setFullYear(newDate.getFullYear() - 1);
							mon = 12 + (mon - count);
							// console.log(`parseDate() |     mon: ${mon}`);
							newDate.setMonth(mon);
						}
						result = newDate.toISOString().substr(0, 7)
						console.log(`parseDate() | month | result: ${result}`);
						break;
					case 'year':
					case 'years':
						result = newDate.getFullYear() - count;
						break;
					default:
						console.error(`parseDate() | unknown unit: '${unit}'`)
					}
				}
				// console.log(`parseDate() |   count: ${count} | unit: ${unit} | year: ${year} | month: ${month} | day: ${day}`);
				// console.log(`parseDate() | newDate: ${newDate.toISOString().substr(0, 10)}`);
				// console.log(`parseDate() |  result: ${result}`);
			}
			// console.log(`parseDate() | result: ${result} | input: "${date}" | test: ${checkDateRE.test(date)}`);
			return result
		}

		switch(true) {
		case D.URL.indexOf('crates.io') !== -1:
			title = D.title.replace('crates.io: Rust Package Registry', 'Crates.io');
			break;
		case D.URL.indexOf('docs.rs') !== -1:
			title = D.title.replace('Rust', 'Docs.rs');
			break;
		case D.URL.indexOf('lib.rs') !== -1:
			title = D.title.replace(' // ', ' - ').replace('—', '-');
			break;
		case D.URL.indexOf('youtube.com/watch') !== -1:
			let channel = D.querySelector('div#owner #text > a').innerText;
			let duration = D.querySelector('#movie_player .ytp-time-duration').innerText;
			let posted = D.querySelector('#description #tooltip').innerText.trim();
			if (posted !== '') {
				posted = posted.split('•')[1].trim();
			} else {
				posted = D.querySelector('#description #info span:nth-child(3)').innerText.trim();
			}
			posted = parseDate(posted);
			title = D.title.replace('YouTube',`${channel} - ${posted} - ${duration} - YouTube`);
			break;
		};

		alert(`[${title.replaceAll('[', "\\[").replaceAll(']', "\\]").replaceAll('|', '&vert;')}](${D.URL})`);
	}
)();

markdown-page-ref-min.js
/* M↓ Page Ref - Version: 1.3.1 */
javascript:(function()%7Bjavascript%3A(%0A%09()%20%3D%3E%20%7B%0A%09%09let%20D%3Ddocument%2C%20title%3DD.title%3B%0A%0A%09%09function%20parseDate(date)%20%7B%0A%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20%3E%20input%3A%20%22%24%7Bdate%7D%22%20%7C%20test%3A%20%24%7BcheckDateRE.test(date)%7D%60)%3B%0A%09%09%09const%20checkDateRE%20%3D%20%2F%5E(%3F%3Cprefix%3E%5BPp%5Dremiered%3F)%3F%20%3F((%3F%3Ccount%3E%5Cd%2B)%20(%3F%3Cunit%3E%5Cw%2B)%20(ago)%7C(%3F%3Cmonth%3E%5Cw%2B)%20(%3F%3Cday%3E%5Cd%2B)%2C%20(%3F%3Cyear%3E%5Cd%2B))%24%2F%3B%0A%09%09%09let%20result%20%3D%20''%3B%0A%0A%09%09%09if%20(checkDateRE.test(date))%20%7B%0A%09%09%09%09let%20%7Bprefix%2C%20count%2C%20unit%2C%20month%2C%20day%2C%20year%7D%20%3D%20checkDateRE.exec(date).groups%3B%0A%09%09%09%09let%20newDate%20%3D%20new%20Date(Date.now())%3B%0A%09%09%09%09let%20transDay%20%3D%20new%20Date(newDate.getFullYear()%2C%20newDate.getMonth()%2C%20newDate.getDate())%3B%0A%09%09%09%09let%20transMonth%20%3D%20new%20Date(newDate.getFullYear()%2C%20newDate.getMonth())%3B%0A%09%09%09%09let%20transYear%20%3D%20new%20Date(newDate.getFullYear()%2C%200%2C%201)%3B%0A%0A%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20%20prefix%3A%20%24%7Bprefix%7D%20%7C%20count%3A%20%24%7Bcount%7D%20%7C%20unit%3A%20%24%7Bunit%7D%20%7C%20month%3A%20%24%7Bmonth%7D%20%7C%20day%3A%20%24%7Bday%7D%20%7C%20year%3A%20%24%7Byear%7D%60)%3B%0A%0A%09%09%09%09if%20(year%20!%3D%20null%20%26%26%20month%20!%3D%20null%20%26%26%20day%20!%3D%20null)%20%7B%0A%09%09%09%09%09console.log(%60parseDate()%20%7C%20Input%20Date%20Normal%20%7C%20%24%7Bdate%7D%60)%3B%0A%09%09%09%09%09newDate%20%3D%20new%20Date(date)%3B%0A%09%09%09%09%09result%20%3D%20newDate.toISOString().substr(0%2C%2010)%3B%0A%09%09%09%09%7D%20else%20if%20(count%20!%3D%20null%20%26%26%20unit%20!%3D%20null)%20%7B%0A%09%09%09%09%09let%20days%20%3D%200%3B%0A%0A%09%09%09%09%09console.log(%60parseDate()%20%7C%20Input%20Type%20Relative%20%7C%20%24%7Bdate%7D%60)%3B%0A%09%09%09%09%09unit%20%3D%20unit.toLowerCase()%3B%0A%09%09%09%09%09console.log(%60parseDate()%20%7C%20%20%20%20unit%3A%20%24%7Bunit%7D%60)%3B%0A%09%09%09%09%09switch%20(unit)%20%7B%0A%09%09%09%09%09case%20'second'%3A%0A%09%09%09%09%09case%20'seconds'%3A%0A%09%09%09%09%09%09if%20(newDate.getTime()%20%3C%20transYear.getTime())%20%7B%0A%09%09%09%09%09%09%09newDate.setFullYear(newDate.getFullYear()%20-%201)%3B%0A%09%09%09%09%09%09%7D%0A%09%09%09%09%09%09%2F%2F%20break%3B%0A%09%09%09%09%09case%20'minute'%3A%0A%09%09%09%09%09case%20'minutes'%3A%0A%09%09%09%09%09case%20'hour'%3A%0A%09%09%09%09%09case%20'hours'%3A%0A%09%09%09%09%09%09%2F%2F%20Our%20job%20is%20done%0A%09%09%09%09%09%09result%20%3D%20newDate.toISOString().substr(0%2C%2010)%0A%09%09%09%09%09%09break%3B%0A%09%09%09%09%09case%20'day'%3A%0A%09%09%09%09%09case%20'days'%3A%0A%09%09%09%09%09case%20'week'%3A%0A%09%09%09%09%09case%20'weeks'%3A%0A%09%09%09%09%09%09newDate.setHours(0)%3B%0A%09%09%09%09%09%09newDate.setMinutes(0)%3B%0A%09%09%09%09%09%09newDate.setSeconds(0)%3B%0A%09%09%09%09%09%09if%20(unit%20%3D%3D%3D%20'week'%20%7C%7C%20unit%20%3D%3D%3D%20'weeks')%20%7B%0A%09%09%09%09%09%09%09count%20%3D%20count%20*%207%3B%0A%09%09%09%09%09%09%7D%0A%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%200%20%7C%20newDate.toISOString().substr(0%2C%2010)%3A%20%24%7BnewDate.toISOString().substr(0%2C%2010)%7D%60)%3B%0A%09%09%09%09%09%09days%20%3D%20newDate.getDate()%3B%0A%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%200%20%7C%20days%3A%20%24%7Bdays%7D%20%7C%20count%3A%20%24%7Bcount%7D%60)%3B%0A%09%09%09%09%09%09if%20(count%20%3C%20days)%20%7B%0A%09%09%09%09%09%09%09days%20-%3D%20count%3B%0A%09%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%201%20%7C%20days%3A%20%24%7Bdays%7D%20%7C%20count%3A%20%24%7Bcount%7D%60)%3B%0A%09%09%09%09%09%09%09newDate.setDate(days)%3B%0A%09%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%202%20%7C%20days%3A%20%24%7BnewDate.getDate()%7D%20%7C%20count%3A%20%24%7Bcount%7D%60)%3B%0A%09%09%09%09%09%09%7D%20else%20%7B%0A%09%09%09%09%09%09%09let%20mon%20%3D%20newDate.getMonth()%3B%0A%09%09%09%09%09%09%09if%20(mon%20%3D%3D%3D%200)%20%7B%0A%09%09%09%09%09%09%09%09newDate.setYear(newDate.getFullYear()%20-%201)%3B%0A%09%09%09%09%09%09%09%09newDate.setMonth(11)%3B%0A%09%09%09%09%09%09%09%7D%20else%20%7B%0A%09%09%09%09%09%09%09%09newDate.setMonth(mon%20-%201)%3B%0A%09%09%09%09%09%09%09%7D%0A%09%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20%20%20%20%20mon%3A%20%24%7Bmon%7D%60)%3B%0A%09%09%09%09%09%09%7D%0A%09%09%09%09%09%09if%20(unit%20%3D%3D%3D%20'week'%20%7C%7C%20unit%20%3D%3D%3D%20'weeks')%20%7B%0A%09%09%09%09%09%09%09result%20%3D%20newDate.toISOString().substr(0%2C%207)%0A%09%09%09%09%09%09%7D%20else%20%7B%0A%09%09%09%09%09%09%09result%20%3D%20newDate.toISOString().substr(0%2C%2010)%0A%09%09%09%09%09%09%7D%0A%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%203%20%7C%20days%3A%20%24%7BnewDate.getDate()%7D%20%7C%20result%3A%20%24%7Bresult%7D%60)%3B%0A%09%09%09%09%09%09break%3B%0A%09%09%09%09%09case%20'month'%3A%0A%09%09%09%09%09case%20'months'%3A%0A%09%09%09%09%09%09newDate.setDate(1)%0A%09%09%09%09%09%09newDate.setHours(0)%3B%0A%09%09%09%09%09%09newDate.setMinutes(0)%3B%0A%09%09%09%09%09%09newDate.setSeconds(0)%3B%0A%0A%09%09%09%09%09%09let%20mon%20%3D%20newDate.getMonth()%3B%0A%09%09%09%09%09%09console.log(%60parseDate()%20%7C%20%20%20%20%20mon%3A%20%24%7Bmon%7D%20%7C%20count%3A%20%24%7Bcount%7D%60)%3B%0A%09%09%09%09%09%09if%20(count%20%3C%20mon)%20%7B%0A%09%09%09%09%09%09%09newDate.setMonth(mon%20-%20count)%3B%0A%09%09%09%09%09%09%7D%20else%20%7B%0A%09%09%09%09%09%09%09newDate.setFullYear(newDate.getFullYear()%20-%201)%3B%0A%09%09%09%09%09%09%09mon%20%3D%2012%20%2B%20(mon%20-%20count)%3B%0A%09%09%09%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20%20%20%20%20mon%3A%20%24%7Bmon%7D%60)%3B%0A%09%09%09%09%09%09%09newDate.setMonth(mon)%3B%0A%09%09%09%09%09%09%7D%0A%09%09%09%09%09%09result%20%3D%20newDate.toISOString().substr(0%2C%207)%0A%09%09%09%09%09%09console.log(%60parseDate()%20%7C%20month%20%7C%20result%3A%20%24%7Bresult%7D%60)%3B%0A%09%09%09%09%09%09break%3B%0A%09%09%09%09%09case%20'year'%3A%0A%09%09%09%09%09case%20'years'%3A%0A%09%09%09%09%09%09result%20%3D%20newDate.getFullYear()%20-%20count%3B%0A%09%09%09%09%09%09break%3B%0A%09%09%09%09%09default%3A%0A%09%09%09%09%09%09console.error(%60parseDate()%20%7C%20unknown%20unit%3A%20'%24%7Bunit%7D'%60)%0A%09%09%09%09%09%7D%0A%09%09%09%09%7D%0A%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20%20%20count%3A%20%24%7Bcount%7D%20%7C%20unit%3A%20%24%7Bunit%7D%20%7C%20year%3A%20%24%7Byear%7D%20%7C%20month%3A%20%24%7Bmonth%7D%20%7C%20day%3A%20%24%7Bday%7D%60)%3B%0A%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20newDate%3A%20%24%7BnewDate.toISOString().substr(0%2C%2010)%7D%60)%3B%0A%09%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20%20result%3A%20%24%7Bresult%7D%60)%3B%0A%09%09%09%7D%0A%09%09%09%2F%2F%20console.log(%60parseDate()%20%7C%20result%3A%20%24%7Bresult%7D%20%7C%20input%3A%20%22%24%7Bdate%7D%22%20%7C%20test%3A%20%24%7BcheckDateRE.test(date)%7D%60)%3B%0A%09%09%09return%20result%0A%09%09%7D%0A%0A%09%09switch(true)%20%7B%0A%09%09case%20D.URL.indexOf('crates.io')%20!%3D%3D%20-1%3A%0A%09%09%09title%20%3D%20D.title.replace('crates.io%3A%20Rust%20Package%20Registry'%2C%20'Crates.io')%3B%0A%09%09%09break%3B%0A%09%09case%20D.URL.indexOf('docs.rs')%20!%3D%3D%20-1%3A%0A%09%09%09title%20%3D%20D.title.replace('Rust'%2C%20'Docs.rs')%3B%0A%09%09%09break%3B%0A%09%09case%20D.URL.indexOf('lib.rs')%20!%3D%3D%20-1%3A%0A%09%09%09title%20%3D%20D.title.replace('%20%2F%2F%20'%2C%20'%20-%20').replace('%E2%80%94'%2C%20'-')%3B%0A%09%09%09break%3B%0A%09%09case%20D.URL.indexOf('youtube.com%2Fwatch')%20!%3D%3D%20-1%3A%0A%09%09%09let%20channel%20%3D%20D.querySelector('div%23owner%20%23text%20%3E%20a').innerText%3B%0A%09%09%09let%20duration%20%3D%20D.querySelector('%23movie_player%20.ytp-time-duration').innerText%3B%0A%09%09%09let%20posted%20%3D%20D.querySelector('%23description%20%23tooltip').innerText.trim()%3B%0A%09%09%09if%20(posted%20!%3D%3D%20'')%20%7B%0A%09%09%09%09posted%20%3D%20posted.split('%E2%80%A2')%5B1%5D.trim()%3B%0A%09%09%09%7D%20else%20%7B%0A%09%09%09%09posted%20%3D%20D.querySelector('%23description%20%23info%20span%3Anth-child(3)').innerText.trim()%3B%0A%09%09%09%7D%0A%09%09%09posted%20%3D%20parseDate(posted)%3B%0A%09%09%09title%20%3D%20D.title.replace('YouTube'%2C%60%24%7Bchannel%7D%20-%20%24%7Bposted%7D%20-%20%24%7Bduration%7D%20-%20YouTube%60)%3B%0A%09%09%09break%3B%0A%09%09%7D%3B%0A%0A%09%09alert(%60%5B%24%7Btitle.replaceAll('%5B'%2C%20%22%5C%5C%5B%22).replaceAll('%5D'%2C%20%22%5C%5C%5D%22).replaceAll('%7C'%2C%20'%26vert%3B')%7D%5D%3A%20%24%7BD.URL%7D%60)%3B%0A%09%7D%0A)()%3B%7D)()%3B

markdown-selection-ref.js
/*
M↓ Selection Ref

Version: 1.3.1
Description: YouTube and Rust community aware
Created: 2023-04-05
Updated: 2023-10-24
*/
javascript:(
	() => {
		let D=document, title=window.getSelection().toString()+' - ';

		function parseDate(date) {
			// console.log(`parseDate() | > input: "${date}" | test: ${checkDateRE.test(date)}`);
			const checkDateRE = /^(?<prefix>[Pp]remiered?)? ?((?<count>\d+) (?<unit>\w+) (ago)|(?<month>\w+) (?<day>\d+), (?<year>\d+))$/;
			let result = '';

			if (checkDateRE.test(date)) {
				let {prefix, count, unit, month, day, year} = checkDateRE.exec(date).groups;
				let newDate = new Date(Date.now());
				let transDay = new Date(newDate.getFullYear(), newDate.getMonth(), newDate.getDate());
				let transMonth = new Date(newDate.getFullYear(), newDate.getMonth());
				let transYear = new Date(newDate.getFullYear(), 0, 1);

				// console.log(`parseDate() |  prefix: ${prefix} | count: ${count} | unit: ${unit} | month: ${month} | day: ${day} | year: ${year}`);

				if (year != null && month != null && day != null) {
					console.log(`parseDate() | Input Date Normal | ${date}`);
					newDate = new Date(date);
					result = newDate.toISOString().substr(0, 10);
				} else if (count != null && unit != null) {
					let days = 0;

					console.log(`parseDate() | Input Type Relative | ${date}`);
					unit = unit.toLowerCase();
					console.log(`parseDate() |    unit: ${unit}`);
					switch (unit) {
					case 'second':
					case 'seconds':
						if (newDate.getTime() < transYear.getTime()) {
							newDate.setFullYear(newDate.getFullYear() - 1);
						}
						// break;
					case 'minute':
					case 'minutes':
					case 'hour':
					case 'hours':
						// Our job is done
						result = newDate.toISOString().substr(0, 10)
						break;
					case 'day':
					case 'days':
					case 'week':
					case 'weeks':
						newDate.setHours(0);
						newDate.setMinutes(0);
						newDate.setSeconds(0);
						if (unit === 'week' || unit === 'weeks') {
							count = count * 7;
						}
						// console.log(`parseDate() | 0 | newDate.toISOString().substr(0, 10): ${newDate.toISOString().substr(0, 10)}`);
						days = newDate.getDate();
						// console.log(`parseDate() | 0 | days: ${days} | count: ${count}`);
						if (count < days) {
							days -= count;
							// console.log(`parseDate() | 1 | days: ${days} | count: ${count}`);
							newDate.setDate(days);
							// console.log(`parseDate() | 2 | days: ${newDate.getDate()} | count: ${count}`);
						} else {
							let mon = newDate.getMonth();
							if (mon === 0) {
								newDate.setYear(newDate.getFullYear() - 1);
								newDate.setMonth(11);
							} else {
								newDate.setMonth(mon - 1);
							}
							// console.log(`parseDate() |     mon: ${mon}`);
						}
						if (unit === 'week' || unit === 'weeks') {
							result = newDate.toISOString().substr(0, 7)
						} else {
							result = newDate.toISOString().substr(0, 10)
						}
						// console.log(`parseDate() | 3 | days: ${newDate.getDate()} | result: ${result}`);
						break;
					case 'month':
					case 'months':
						newDate.setDate(1)
						newDate.setHours(0);
						newDate.setMinutes(0);
						newDate.setSeconds(0);

						let mon = newDate.getMonth();
						console.log(`parseDate() |     mon: ${mon} | count: ${count}`);
						if (count < mon) {
							newDate.setMonth(mon - count);
						} else {
							newDate.setFullYear(newDate.getFullYear() - 1);
							mon = 12 + (mon - count);
							// console.log(`parseDate() |     mon: ${mon}`);
							newDate.setMonth(mon);
						}
						result = newDate.toISOString().substr(0, 7)
						console.log(`parseDate() | month | result: ${result}`);
						break;
					case 'year':
					case 'years':
						result = newDate.getFullYear() - count;
						break;
					default:
						console.error(`parseDate() | unknown unit: '${unit}'`)
					}
				}
				// console.log(`parseDate() |   count: ${count} | unit: ${unit} | year: ${year} | month: ${month} | day: ${day}`);
				// console.log(`parseDate() | newDate: ${newDate.toISOString().substr(0, 10)}`);
				// console.log(`parseDate() |  result: ${result}`);
			}
			// console.log(`parseDate() | result: ${result} | input: "${date}" | test: ${checkDateRE.test(date)}`);
			return result
		}

		switch(true) {
		case D.URL.indexOf('crates.io') !== -1:
			title += D.title.replace('crates.io: Rust Package Registry', 'Crates.io');
			break;
		case D.URL.indexOf('docs.rs') !== -1:
			title += D.title.replace('Rust', 'Docs.rs');
			break;
		case D.URL.indexOf('lib.rs') !== -1:
			title += D.title.replace(' // ', ' - ').replace('—', '-');
			break;
		case D.URL.indexOf('youtube.com/watch') !== -1:
			let channel = D.querySelector('div#owner #text > a').innerText;
			let duration = D.querySelector('#movie_player .ytp-time-duration').innerText;
			let posted = D.querySelector('#description #tooltip').innerText.trim();
			if (posted !== '') {
				posted = posted.split('•')[1].trim();
			} else {
				posted = D.querySelector('#description #info span:nth-child(3)').innerText.trim();
			}
			posted = parseDate(posted);
			title += D.title.replace('YouTube',`${channel} - ${posted} - ${duration} - YouTube`);
			break;
		default:
			title += D.title;
			break;
		};

		alert(`[${title.replaceAll('[', "\\[").replaceAll(']', "\\]").replaceAll('|', '&vert;')}]: ${D.URL}`);
	}
)();